File: Algorithm Program/22-23/fall/lecture/math/3-sieve-of-eratosthenes/sieve.py
# https://ideone.com/7CYqnz
n = 10**6


def sieve(n: int) -> "list[bool]":
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = 1
    for i in range(2, n + 1):
        for multiple in range(2 * i, n + 1, i):
            is_prime[multiple] = False
    return is_prime


is_prime = sieve(n)
n_primes = 0
for i in range(2, n + 1):
    n_primes += is_prime[i]

print("Number of primes up to", n, "=", n_primes)

File: Algorithm Program/22-23/fall/lecture/math/3-sieve-of-eratosthenes/sieve.cpp
// https://ideone.com/NyqgrL
#include <iostream>
#include <vector>
using namespace std;

vector<bool> sieve(int n) {
    // Initially, every integer is marked as prime.
    // By convention, we named the array "is_prime" instead of "is_composite".
    vector<bool> is_prime(n + 1, true);
    is_prime[0] = is_prime[1] = false; // Should mark them by hand

    // Now, start marking composites
    for (int i = 2; i <= n; i++) {
        for (int multiple = 2 * i; multiple <= n; multiple += i) { // n/i iterations
            is_prime[multiple] = false;
        }
    }
    return is_prime;
}

int main() {
    int n = 1e6;
    vector<bool> is_prime = sieve(n);
    int n_primes = 0;
    for (int i = 2; i <= n; i++) {
        n_primes += is_prime[i];
    }
    cout << "Number of primes up to " << n << " = " << n_primes << "!\n";
}